homeassistant/components/light/mqtt.py

```diff
@@ -0,0 +1,171 @@
+        self._state_topic = state_topic
+        self._command_topic = command_topic
+        self._brightness_state_topic = brightness_state_topic
+        self._brightness_command_topic = brightness_command_topic
+        self._rgb_state_topic = rgb_state_topic
+        self._rgb_command_topic = rgb_command_topic
+        self._rgb = rgb
+        self._qos = qos
+        self._payload_on = payload_on
+        self._payload_off = payload_off
+        self._brightness = brightness
+        self._xy = [[0.5, 0.5]]
+        self._state = False
+
+        def message_received(topic, payload, qos):
+            """ A new MQTT message has been received. """
+            if payload == self._payload_on:
+                self._state = True
+                self.update_ha_state()
+            elif payload == self._payload_off:
+                self._state = False
+                self.update_ha_state()
+
+        def brightness_received(topic, payload, qos):
+            """ A new MQTT message has been received. """
+            self._brightness = int(payload)
+            self.update_ha_state()
+
+        def rgb_received(topic, payload, qos):
+            """ A new MQTT message has been received. """
+            rgb = payload.split(",")
+            self._rgb = list(map(int, rgb))
+            self.update_ha_state()
+
+        # subscribe the state_topic
+        mqtt.subscribe(self._hass, self._state_topic,
+                       message_received, self._qos)
+        mqtt.subscribe(self._hass, self._brightness_state_topic,
+                       brightness_received, self._qos)
+        mqtt.subscribe(self._hass, self._rgb_state_topic,
+                       rgb_received, self._qos)
+
+    @property
+    def should_poll(self):
+        """ No polling needed for a demo light. """
+        return False
+
+    @property
+    def name(self):
+        """ Returns the name of the device if any. """
+        return self._name
+
+    @property
+    def brightness(self):
+        """ Brightness of this light between 0..255. """
+        return self._brightness
+
+    @property
+    def rgb_color(self):
+        """ RGB color value. """
+        return self._rgb
+
+    @property
+    def color_xy(self):
+        """ RGB color value. """
+        return self._xy
+
+    @property
+    def is_on(self):
+        """ True if device is on. """
+        return self._state
+
+    def turn_on(self, **kwargs):
+        """ Turn the device on. """
+
+        if ATTR_RGB_COLOR in kwargs:
+            self._rgb = kwargs[ATTR_RGB_COLOR]
+            rgb = DEFAULT_RGB_PATTERN % tuple(self._rgb)
+            mqtt.publish(self._hass, self._rgb_command_topic, rgb, self._qos)
+
+        if ATTR_BRIGHTNESS in kwargs:
+            self._brightness = kwargs[ATTR_BRIGHTNESS]
+            mqtt.publish(self._hass, self._brightness_command_topic,
+                         self._brightness, self._qos)
+
+        if not self._state:
+            self._state = True
```

Comment:
This is optimistic behavior. The light should only act optimistic if there is no state topic defined. If we have a state topic, the new state will come to us via the state topic.